typing/shape.ml

```diff
@@ -336,13 +372,20 @@ end) = struct
      same hash.
 *)
 
-  and reduce__ ({fuel; global_env; local_env; _} as env) (t : t) =
-    let reduce env t = reduce_ env t in
-    let delay_reduce env t = Thunk (env.local_env, t) in
-    let force (Thunk (local_env, t)) =
+  and reduce__
+    ({fuel; global_env; local_env; _} as env) (t : t) =
+    let reduce env t =
+      reduce_ env t
+    in
+    let delay_reduce { local_env; _ } t =
+      Thunk { local_env; shape = t }
+    in
+    let force (Thunk { local_env; shape = t }) =
       reduce { env with local_env } t in
-    let return desc : nf = { uid = t.uid; desc } in
-    if !fuel < 0 then return (NoFuelLeft t.desc)
+    let return ?(approximated = t.approximated) desc : nf =
+      { uid = t.uid; desc; approximated }
+    in
+    if !fuel < 0 then return ~approximated:true (NoFuelLeft t.desc)
```

Comment:
(putting this comment here, even if the problem existed before this PR, since my other comment made me think about that)
Here, we have again a case where we return only `t`, while ignoring its environment. We might actually want to substitute a weak read back of the environment in `t` here to avoid the binding problems.